jep/0000/README.adoc

```diff
@@ -0,0 +1,225 @@
+
+== Specification
+
+Jenkins agents upload and download artifacts directly to the external location using HTTP, no content traffic is sent over the remoting channel.
+
+All external store operations except upload/download are executed from the master, so the agent does not need any permissions other than HTTP upload/download. Pre-signed urls with expiration can be used so the agent does not have access to the full store.
+
+Stashes are stored as tarballs to ensure permissions and links are preserved.
+
+Any operations using `ArtifactManager` remain unaffected and will transparently use the chosen implementation.
+
+The flow looks like
+
+* Upload:
+** master creates blob metadata in external location
+** master gets pre-signed url for blob upload operation
+** url is sent to agent
+** agent creates a tarball in case of stashing
+** agent does the actual file upload through HTTP
+* Download:
+** master gets pre-signed url for blob download operation
+** url is sent to agent
+** agent does the actual file download through HTTP
+** agent unpacks the tarball in case of unstashing
+* Browser:
+** user clicks on artifact download link
+** master gets pre-signed url for blob download operation
+** browser is sent a HTTP redirect to that url so download is between external store and browser
+
+
+This change adds more methods to the VirtualFile API to support this flow.
+
+The `VirtualFile.run(Callable)` method, that already existed to optimize Workspace link browsing on agents, is now being used in artifact-related code in Jenkins core.
+It was important to implement this in the S3 plugin in order to ensure that Classic or Blue Ocean UI flows relating to artifacts, including simply opening the index page of a build, did not make a large number of network calls.
+The cache system allows a single S3 metadata call (possibly one HTTP request, depending on page size) to retrieve sufficient information for a complete page render of some or all of the artifact tree, without the rendering code (e.g., in Jenkins core) knowing anything about the storage system.
+
+
+== Motivation
+
+As mentioned in the abstract, running in a containerized environment is challenging due to the excessive use of the file system. Plus the lack of functionality provided by files vs dedicated artifact or blob store (versioning, permissions, cold store,...)
+
+From the usability point of view, currently a Jenkins installation sizing has to have into account the expected amount of disk needed or a restart, disk reprovisioning, disk swap,... instead of a infinite scale, pay per use that a blob store can offer. Cloud installations can automatically configure the blob store to use offering a much better first time user experience.
+
+Several alternatives exist today but they all require changes to all the pipelines and job definitions to explicitly choose the backend to send artifacts to.
+
+* https://plugins.jenkins.io/s3[S3 plugin]
```

Comment:
The plugins should be able to co-exist well if needed, right? Also CC @Jimilian

Reply:
yes, it wouldn't affect it